include/alpaka/mem/buf/sycl/Copy.hpp

```diff
@@ -21,129 +21,367 @@
+        std::uint8_t* const m_dstMemNative;
+        std::uint8_t const* const m_srcMemNative;
     };
 
-    template<typename TSrc, typename TDst, Direction TDirection>
-    struct TaskCopySycl
+    //! The Sycl device ND memory copy task.
+    template<typename TDim, typename TViewDst, typename TViewSrc, typename TExtent>
+    struct TaskCopySycl : public TaskCopySyclBase<TDim, TViewDst, TViewSrc, TExtent>
     {
-        auto operator()(sycl::handler& cgh) const -> void
+        using DimMin1 = DimInt<TDim::value - 1u>;
+        using typename TaskCopySyclBase<TDim, TViewDst, TViewSrc, TExtent>::ExtentSize;
+        using typename TaskCopySyclBase<TDim, TViewDst, TViewSrc, TExtent>::DstSize;
+        using typename TaskCopySyclBase<TDim, TViewDst, TViewSrc, TExtent>::SrcSize;
+
+        using TaskCopySyclBase<TDim, TViewDst, TViewSrc, TExtent>::TaskCopySyclBase;
+
+        template<typename TQueue>
+        ALPAKA_FN_HOST auto enqueue(TQueue& queue) const -> void
         {
-            if constexpr(TDirection == Direction::d2h || TDirection == Direction::d2d)
-                cgh.require(m_src);
+            ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
 
-            if constexpr(TDirection == Direction::h2d || TDirection == Direction::d2d)
-                cgh.require(m_dst);
+#    if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL
+            this->printDebug();
+#    endif
+            // [z, y, x] -> [z, y] because all elements with the innermost x dimension are handled within one
+            // iteration.
+            Vec<DimMin1, ExtentSize> const extentWithoutInnermost(subVecBegin<DimMin1>(this->m_extent));
+            // [z, y, x] -> [y, x] because the z pitch (the full size of the buffer) is not required.
+            Vec<DimMin1, DstSize> const dstPitchBytesWithoutOutmost(subVecEnd<DimMin1>(this->m_dstPitchBytes));
+            Vec<DimMin1, SrcSize> const srcPitchBytesWithoutOutmost(subVecEnd<DimMin1>(this->m_srcPitchBytes));
 
-            cgh.copy(m_src, m_dst);
+            if(static_cast<std::size_t>(this->m_extent.prod()) != 0u)
+            {
+                meta::ndLoopIncIdx(
+                    extentWithoutInnermost,
+                    [&](Vec<DimMin1, ExtentSize> const& idx)
+                    {
+                        queue.getNativeHandle().memcpy(
+                            reinterpret_cast<void*>(
+                                this->m_dstMemNative
+                                + (castVec<DstSize>(idx) * dstPitchBytesWithoutOutmost)
+                                      .foldrAll(std::plus<DstSize>())),
+                            reinterpret_cast<void const*>(
+                                this->m_srcMemNative
+                                + (castVec<SrcSize>(idx) * srcPitchBytesWithoutOutmost)
+                                      .foldrAll(std::plus<SrcSize>())),
+                            static_cast<std::size_t>(this->m_extentWidthBytes));
+                    });
+            }
```

Comment:
This would launch many `memcpy` operations if we are dealing with 2D or 3D buffers. I assume you are doing this in order to deal with offsets / slices in one of the source / destination views. Wouldn't it be more effective to write a specialized copy kernel for the 2D / 3D case?

Reply:
does Alpaka support the cases where
  - the source and destination buffers have different pitch
  - only a subset of the source and destination buffers are copied 

?

Reply:
At least the CUDA implementation seems to support both points, so I'd argue for SYCL to do the same. CC @psychocoderHPC, please correct me if I'm wrong.